### PR TITLE
[doc] Improve tutorial on exporting results

### LINKED ISSUE #1176: [Doc] [GUI] improving the export_results doc and make imwrite have the same behavior with ti.GUI.show(filename)
**Concisely describe the proposed feature**
As @archibate mentioned before, I want to improve the docs for these 2 points:
1. More explanations for the datatypes used in `ti.imwrite`, note that `ti.f32` can also be used to export images.
2. Add some examples which use `ti.Matrix` and `ti.Vector` with `ti.imwrite()`, so that our user's mind won't be restricted in `ti.var`

**Besides, I also find that in our newest release**, `ti.imwrite` is changed to:

https://github.com/taichi-dev/taichi/blob/43904488fdaca626178d45b467aba6989ebe5fe8/python/taichi/misc/image.py#L5-L20

In these codes above, we first assert:
`assert len(img.shape) in [2, 3], "Image must be either RGB/RGBA or greyscale"`

But If the length of `image.shape` is really two, our second assertation will fail by out-of-bounds accessing for `img.shape`:

`assert img.shape[2] in [1, 3, 4], "Image must be either RGB/RGBA or greyscale" `

Maybe I can help to fix it.

**A clear and concise description of what you want.**

1. I will add some more examples and explanations in export_result.rst and gui.rst

2. Maybe I can help to do some minor revision in `ti.imwrite`

**Describe the solution you'd like (if any)**
in image.py, change the code so that we only have the second assertation (mentioned above) when `len(img.shape) == 3`.

**Additional comments**
Nothing more.

### PR DESCRIPTION
**Pull request by Rullec**
Related issue = Close #1176 

Hi all,

  This PR tries to offer more examples and explantions in export_results.rst and gui.rst.

  I have proofread x5 and think about this topic x5. But I guess there must still be some mistakes and unsatisfied stuff :D. If so, please let me know.

  Hopefully this PR can make our docs better!

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: docs/export_results.rst [hunk 1]
```diff
@@ -68,7 +67,21 @@ To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
         ti.imwrite(pixels.to_numpy(), filename)
         print(f'The image has been saved to {filename}')
 
-Taichi offers helper functions that read, write, and show images. Please check out :ref:`gui` for more details.
+- ``ti.imwrite`` can export both Taichi tensors ``ti.Matrix``, ``ti.Vector`` or ``ti.var`` and numpy tensors ``np.ndarray``. 
+- Same as above ``ti.GUI.show(filename)``, the image format (``png``, ``jpg`` and ``bmp``) is also controlled by the suffix of ``filename`` in ``ti.imwrite(filename)``.
+- Meanwhile, The resulted image type (``grayscale``, ``RGB``, or ``RGBA``) is determined by **the number of channels in the input tensor**, aka the length of the third dimension, i.e. ``tensor.shape[2]``.
```
Comment (archibate) [inline docs/export_results.rst:72]:
```suggestion
- Meanwhile, the resulted image type (grayscale, RGB, or RGBA) is determined by **the number of channels in the input tensor**, aka the length of the third dimension, i.e. ``tensor.shape[2]``. 
```

File: docs/export_results.rst [hunk 2]
```diff
@@ -47,8 +47,7 @@ Export images using ``ti.GUI.show``
 
 Export images using ``ti.imwrite``
 ++++++++++++++++++++++++++++++++++
-
-To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
+To save images without creating a ``ti.GUI.show(filename)``, use ``ti.imwrite(filename)``. For example:
```
Comment (archibate) [inline docs/export_results.rst:50]:
```suggestion
To save images without invoking ``ti.GUI.show(filename)``, use ``ti.imwrite(filename)``. For example:
```

File: docs/gui.rst [hunk 3]
```diff
@@ -328,14 +392,4 @@ Image I/O
...
-    :parameter img: (Matrix or Expr) the images you want to export
-    :parameter filename: (string) filename you want to save
-
-    Please make sure that the input image is a Taichi tensor of scalar or Vector, and **the tensor has correct shape (height, width, components)**, see ``ti.GUI.set_image``. For example:
-.. code-block:: python
-
-    img = ti.imread('hello.png')
-    ti.imshow(img, 'Window Title')
-    ti.imwrite(img, 'hello2.png')
+    It has the same logic as ``ti.imwrite`` for differnt datatypes.
```
Comment (archibate) [inline docs/gui.rst:395]:
```suggestion
    It has the same logic as ``ti.imwrite`` for different datatypes.
```

File: docs/gui.rst [hunk 4]
```diff
@@ -314,12 +314,76 @@ A *event filter* is a list combined of *key*, *type* and *(type, key)* tuple, e.
...
 .. function:: ti.imread(filename, channels=0)
 
     :parameter filename: (string) the filename of the image to load
-    :parameter channels: (int) the number of channels in your specified image. The default value ``0`` means the channels of the returned image is adaptive to the image file.
+    :parameter channels: (optional int) the number of channels in your specified image. The default value ``0`` means the channels of the returned image is adaptive to the image file.
+
+    :return: (np.ndarray) the image read from ``filename``
 
-    This function loads an image from the target filename and returns it as a ``np.ndarray``.
+    This function loads an image from the target filename and returns it as a ``np.ndarray(dt=np.uint8)``.
```
Comment (archibate) [inline docs/gui.rst:384]:
```suggestion
    This function loads an image from the target filename and returns it as a ``np.ndarray(dtype=np.uint8)``. 
```

File: docs/gui.rst [hunk 5]
```diff
@@ -314,12 +314,76 @@ A *event filter* is a list combined of *key*, *type* and *(type, key)* tuple, e.
 Image I/O
 ---------
 
+.. function:: ti.imwrite(img, filename)
+
+    :parameter img: (Matrix or Expr) the images you want to export. 
+    :parameter filename: (string) the location you want to save to
+
+    Export a ``np.ndarray`` or Taichi tensor(``ti.Matrix``, ``ti.Vector``, or ``ti.var``) to a specified location ``filename``.
```
Comment (archibate) [inline docs/gui.rst:322]:
```suggestion
    Export a ``np.ndarray`` or Taichi tensor (``ti.Matrix``, ``ti.Vector``, or ``ti.var``) to a specified location ``filename``.
```

File: docs/export_results.rst [hunk 6]
```diff
@@ -68,7 +67,21 @@ To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
...
+- For example, A tensor that has shape ``[w, h]`` or ``[w, h, 1]`` will be exported as a ``grayscale`` image.
+- If you want to export ``RGB`` or ``RGBA`` images instead, the input tensor should have a shape ``[w, h, 3]`` or ``[w, h, 4]`` respectively.
+
+.. note::
+    
+    A Taichi tensor can have a different datatype. Different datatype makes ``ti.imwrite`` have different behaviors.
+    
+    For a floating type(ti.f16, ti.f32) tensor, the pixel value on each channel of the image should be **a floating number between 0 and 1**.
+    
+    For an int type (ti.u8, ti.u16, etc) tensor, the values can be arbitrary valid **integer in its own bounds of datatype**.
```
Comment (archibate) [inline docs/export_results.rst:82]: 80－82 duplicate with gui.rst? Adding a link is good enough.

**Comment by Rullec** [comment]
My fault! Thanks for these suggestions!

I don't want to debate anything for my self. But your suggestions has made me realize that, **my problem is still in the workflow.** Grammerly also cannot do everything for me.

1. I should also check the constructor for `np.ndarray` when writing the doc, so that I can realize the keyword argument is `dtype` but not `dt`. Memory is not reliable.

2.  For those `Meanwhile, The...` `Taichi tensor[ADD A SPACE HERE](``ti.Matrix``)` cases. I really don't know them before... I will take care of these 2 rules next time.

Sincerely sorry about these mistakes!

File: docs/export_results.rst (hunk 6 above)
Comment (Rullec) [reply docs/export_results.rst:82]: Yes, I thought this question when I was writing indeed.
There has been an ref of gui.rst in L84. Will it be a little verbose if I ref it again in L82? How to deal with it properly?

File: docs/gui.rst (hunk 5 above)
Comment (Rullec) [reply docs/gui.rst:322]: Great, a space before brackets.

**Comment by Rullec** [comment]
Update:
1. fix some typos.
2. add simple comments in our example.
3. uniform the format and symbol between ``gui.rst`` and ``export_results.rst``.
4. try to make the text more fluent.

**Comment by archibate** [comment]
Thank for polishing the doc! Would you mind also write an very simple example of using `VideoManager` to generate gif and mp4 without comment just like you wrote in doc? It can serve as an infrastructures when students want to export video.

**Comment by Rullec** [comment]
Certainly! I will finish it after work.

**Comment by Rullec** [comment]
Hi @archibate ,

Please let me double check our requirements in case of a big wasting of time.

At this moment a simple example for ``exporting videos`` has been added in doc ``export_results.rst``. 

Do you mean that I should add an ``export_videos.py`` in our `examples/` in your comment above so that our end users can mimic it? Or I should add another section for ``Exporting Videos`` in ``gui.rst``?

**Comment by archibate** [comment]
> Hi @archibate ,
> 
> Please let me double check our requirements in case of a big wasting of time.
> 
> At this moment a simple example for `exporting videos` has been added in doc `export_results.rst`.
> 
> Do you mean that I should add an `export_videos.py` in our `examples/` in your comment above so that our end users can mimic it? Or I should add another section for `Exporting Videos` in `gui.rst`?

Yes, please add `examples/export_videos.py` :) Just copying from the sample code in current doc is OK.

**Comment by Rullec** [comment]
Thank you so much for this clear instruction! 

I have isolated the example code from our doc ``export_results.rst`` to ``examples/export_results.py``. It is tested on my ``0.6.8`` dev mode Taichi, and in my own observation it works well.

File: docs/export_results.rst [hunk 7]
```diff
@@ -68,7 +67,17 @@ To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
         ti.imwrite(pixels.to_numpy(), filename)
         print(f'The image has been saved to {filename}')
 
-Taichi offers helper functions that read, write, and show images. Please check out :ref:`gui` for more details.
+- ``ti.imwrite`` can export both Taichi tensors ``ti.Matrix``, ``ti.Vector`` or ``ti.var`` and numpy tensors ``np.ndarray``.
```
Comment (yuanming-hu) [inline docs/export_results.rst:70]:
```suggestion
- ``ti.imwrite`` can export Taichi tensors (``ti.Matrix``, ``ti.Vector``, ``ti.var``) and numpy tensors ``np.ndarray``.
```

File: docs/export_results.rst [hunk 8]
```diff
@@ -68,7 +67,17 @@ To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
...
+- Same as above ``ti.GUI.show(filename)``, the image format (``png``, ``jpg`` and ``bmp``) is also controlled by the suffix of ``filename`` in ``ti.imwrite(filename)``.
+- Meanwhile, the resulted image type (grayscale, RGB, or RGBA) is determined by **the number of channels in the input tensor**, aka the length of the third dimension, i.e. ``tensor.shape[2]``.
```
Comment (yuanming-hu) [inline docs/export_results.rst:72]:
```suggestion
- Meanwhile, the resulted image type (grayscale, RGB, or RGBA) is determined by **the number of channels in the input tensor**, i.e., the length of the third dimension (``tensor.shape[2]``).
```

File: docs/export_results.rst [hunk 9]
```diff
@@ -68,7 +67,17 @@ To save images without creating a ``ti.GUI``, use ``ti.imwrite``. For example:
...
+- In other words, a tensor that has shape ``(w, h)`` or ``(w, h, 1)`` will be exported as a grayscale image.
+- If you want to export ``RGB`` or ``RGBA`` images instead, the input tensor should have a shape ``(w, h, 3)`` or ``(w, h, 4)`` respectively.
+
+.. note::
+
+    All Taichi tensors have their own datatypes, such as ``ti.u8`` and ``ti.f32``. Different datatype can make ``ti.imwrite`` have different behaviors. Please check out :ref:`gui` for more details.
```
Comment (yuanming-hu) [inline docs/export_results.rst:78]:
```suggestion
    All Taichi tensors have their own data types, such as ``ti.u8`` and ``ti.f32``. Different data types can lead to different behaviors of ``ti.imwrite``. Please check out :ref:`gui` for more details.
```

**Comment by Rullec** [comment]
> Very nice! Thanks. I think this is ready to merge after the minor suggestions are applied.

Thanks for these suggestions! I have updated all. :D

**Pull request merged**